src/lib/FreeC/LiftedIR/Converter/Type.hs

```diff
@@ -0,0 +1,121 @@
+import qualified FreeC.LiftedIR.Syntax         as LIR
+
+-- | Converts the argument types of a function.
+liftFuncArgTypes
+  :: Maybe Int -- ^ Index of the decreasing argument for recursive functions.
+  -> [IR.VarPat] -- ^ The argument types.
+  -> [LIR.Type]
+liftFuncArgTypes = maybe liftNonRecFuncArgTypes liftRecFuncArgTypes
+
+-- | Converts the argument types of a non recursive function using @convertType@.
+liftNonRecFuncArgTypes :: [IR.VarPat] -> [LIR.Type]
+liftNonRecFuncArgTypes = map $ fromJust . liftVarPatType
+
+-- | Converts the argument types of a recursive function using @convertType@.
+--
+--   The outermost argument at the given index is marked decreasing.
+liftRecFuncArgTypes :: Int -> [IR.VarPat] -> [LIR.Type]
+liftRecFuncArgTypes decIndex args =
+  let convArgs                      = map (fromJust . liftVarPatType) args
+      (startArgs, decArg : endArgs) = splitAt decIndex convArgs
+  in  startArgs ++ (markOutermostDecreasing decArg : endArgs)
+
+-- | Lifts the type of an 'IR.VarPat'. If the argument is strict the type itself
+--   isn't lifted into the @Free@ monad.
+liftVarPatType :: IR.VarPat -> Maybe LIR.Type
+liftVarPatType (IR.VarPat _ _ patType strict) =
+  bool liftType liftType' strict <$> patType
+
+-- | Converts a constructor argument using @convertType@.
+--
+--   All occurrences of the constructed type are marked as structurally smaller.
+liftConArgType :: IR.QName -> IR.Type -> LIR.Type
+liftConArgType ident = markAllDec ident . liftType
+
+-------------------------------------------------------------------------------
+-- Translations                                                              --
+-------------------------------------------------------------------------------
+
+-- | Converts a type from IR to lifted IR by lifting it into the @Free@ monad.
+--
+--   This corresponds to the dagger translation for monotypes as described by
+--   Abel et al.
+--
+--   > τ' = Free τ*
+liftType :: IR.Type -> LIR.Type
+liftType = LIR.FreeTypeCon NoSrcSpan . liftType'
+
+-- | Lifts a type from IR to lifted IR by adding the free arguments to
+--   constructors and lifting function types in the @Free@ monad.
+--
+--   This corresponds to the star translation for monotypes as described by
+--   Abel et al.
+--
+liftType' :: IR.Type -> LIR.Type
+liftType' = flip liftTypeApp' []
+
+-- | Like 'liftType'' but accumulates the type arguments in a second parameter.
+--
+--   > (C τ₁ … τₙ)* = C τ₁* … τₙ*
+--   > (τ₁ -> τ₂)* = τ₁' -> τ₂'
+--   > α* = α
+liftTypeApp' :: IR.Type -> [IR.Type] -> LIR.Type
+liftTypeApp' (IR.TypeCon srcSpan name) ts =
+  LIR.TypeCon srcSpan name (map liftType' ts) False
+liftTypeApp' (IR.TypeVar srcSpan name) [] = LIR.TypeVar srcSpan name
+liftTypeApp' (IR.TypeApp _ l r       ) ts = liftTypeApp' l (r : ts)
+liftTypeApp' (IR.FuncType srcSpan l r) [] =
+  LIR.FuncType srcSpan (liftType l) (liftType r)
+liftTypeApp' _ (_ : _) =
+  error "liftTypeApp': Only type constructors can be applied!"
```

Comment:
Due to the `KindCheckPass`, this error should not occur in practice. However, we should probably still use `reportFatal` here (which would require `liftType` to be lifted into the `Converter` monad, though).